fiddy/derivative_check.py

```diff
@@ -161,3 +162,109 @@ def method(self, *args, **kwargs):
+
+            expected_value = []
+            for output_index in np.ndindex(self.output_indices):
+                element = self.expectation[output_index][direction_index]
+                expected_value.append(element)
+            expected_value = np.array(expected_value).reshape(test_value.shape)
+            expected_values.append(expected_value)
+
+        # debug
+        assert len(expected_values) == len(
+            test_values
+        ), "Mismatch of step sizes"
+
+        results_all = []
+        directional_derivative_check_results = []
+        for step_size in range(0, len(expected_values)):
+            approxs_for_param = []
+            grads_for_param = []
+            results = []
+            for diff_index, directional_derivative in enumerate(
+                self.derivative.directional_derivatives
+            ):
+                try:
+                    for grad, approx in zip(
+                        expected_values[diff_index - 1][step_size - 1],
+                        test_values[diff_index - 1][step_size - 1],
+                    ):
+                        approxs_for_param.append(approx)
+                        grads_for_param.append(grad)
+                    fd_range = np.percentile(approxs_for_param, [0, 100])
+                    fd_mean = np.mean(approxs_for_param)
+                    grad_mean = np.mean(grads_for_param)
+                    if not (fd_range[0] <= grad_mean <= fd_range[1]):
+                        if np.any(
+                            [
+                                abs(x - y) > kwargs["atol"]
+                                for i, x in enumerate(approxs_for_param)
+                                for j, y in enumerate(approxs_for_param)
+                                if i != j
+                            ]
+                        ):
+                            fd_range = abs(fd_range[1] - fd_range[0])
+                            if (
+                                abs(grad_mean - fd_mean)
+                                / abs(fd_range + np.finfo(float).eps)
+                            ) > kwargs["rtol"]:
+                                results.append(False)
+                            else:
+                                results.append(False)
+                        else:
+                            results.append(
+                                None
+                            )  # can't judge consistency / questionable grad approxs
+                    else:
+                        fd_range = abs(fd_range[1] - fd_range[0])
+                        if math.isinf(
+                            (fd_range)
+                            or math.isnan(fd_range)
+                            or math.isinf(fd_mean)
+                            or math.isnan(fd_mean)
+                        ):
+                            results.append(None)
+                        else:
+                            results.append(True)
```

Comment:
Hm, small preference to set the result in the `if` statements and only have one append statement at the end.

e.g.:
```suggestion
                            result = True
                        results.append(result)
```